pkg/apis/allocation/v1/gameserverallocation.go

```diff
@@ -82,6 +82,137 @@ type GameServerAllocationSpec struct {
 	MetaPatch MetaPatch `json:"metadata,omitempty"`
 }
 
+// GameServerSelector contains all the filter options for selecting
+// a GameServer for allocation.
+type GameServerSelector struct {
+	metav1.LabelSelector
+	// [Stage:Alpha]
+	// [FeatureFlag:StateAllocationFilter]
+	// +optional
+	GameServerState *agonesv1.GameServerState `json:"gameServerState,omitempty"`
+	// [Stage:Alpha]
+	// [FeatureFlag:PlayerAllocationFilter]
+	// +optional
+	Players *PlayerSelector `json:"players,omitempty"`
+}
+
+// PlayerSelector is the filter options for a GameServer based on player counts
+type PlayerSelector struct {
+	MinAvailable int64 `json:"minAvailable,omitempty"`
+	MaxAvailable int64 `json:"maxAvailable,omitempty"`
+}
+
+// ApplyDefaults applies default values to the PlayerSelector
+func (s *GameServerSelector) ApplyDefaults() {
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState == nil {
+			state := agonesv1.GameServerStateReady
+			s.GameServerState = &state
+		}
+	}
+
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) {
+		if s.Players == nil {
+			s.Players = &PlayerSelector{}
+		}
+	}
+}
+
+// Matches checks to see if a GameServer matches a given GameServerSelector's criteria
+func (s *GameServerSelector) Matches(gs *agonesv1.GameServer) bool {
+
+	// Assume at this point, this has already been run through Validate(), and it can be converted.
+	// We end up running LabelSelectorAsSelector twice for each allocation, but if we store the results of this
+	// function within the GameServerSelector, we can't fuzz the GameServerAllocation as reflect.DeepEqual
+	// will fail due to the unexported field.
+	selector, _ := metav1.LabelSelectorAsSelector(&s.LabelSelector) // nolint: error
+	if selector == nil {
+		return false
+	}
+
+	// first check labels
+	if !selector.Matches(labels.Set(gs.ObjectMeta.Labels)) {
+		return false
+	}
+
+	// then if state is being checked, check state
+	if runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		if s.GameServerState != nil && gs.Status.State != *s.GameServerState {
+			return false
+		}
+	}
+
+	// then if player count is being checked, check that
+	if runtime.FeatureEnabled(runtime.FeaturePlayerAllocationFilter) {
+		// 0 is unlimited number of players
+		if s.Players != nil && gs.Status.Players != nil && s.Players.MaxAvailable != 0 {
+			available := gs.Status.Players.Capacity - gs.Status.Players.Count
+			if !(available >= s.Players.MinAvailable && available <= s.Players.MaxAvailable) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Validate validates that the selection fields have valid values
+func (s *GameServerSelector) Validate(field string, causes []metav1.StatusCause) ([]metav1.StatusCause, bool) {
+	valid := true
+
+	_, err := metav1.LabelSelectorAsSelector(&s.LabelSelector)
```

Comment:
Just highlighting this change: We end up calling `LabelSelectorAsSelector` twice. Once here in `Validate()`, where we check the error and then again in `Matches()`, where we don't, since we did it before.

I think it's worth it because it creates a much cleaner API surface, but let me know your thoughts.

Reply:
lgtm